core/services/keeper/registry_synchronizer.go

```diff
@@ -0,0 +1,239 @@
+package keeper
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated/keeper_registry_wrapper"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/job"
+	"github.com/smartcontractkit/chainlink/core/services/postgres"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+const syncUpkeepQueueSize = 10
+
+func NewRegistrySynchronizer(
+	job job.Job,
+	contract *keeper_registry_wrapper.KeeperRegistry,
+	orm ORM,
+	syncInterval time.Duration,
+) *RegistrySynchronizer {
+	return &RegistrySynchronizer{
+		contract:      contract,
+		doneWG:        sync.WaitGroup{},
+		orm:           orm,
+		job:           job,
+		interval:      syncInterval,
+		chDone:        make(chan struct{}),
+		StartStopOnce: utils.StartStopOnce{},
+	}
+}
+
+// RegistrySynchronizer conforms to the job.Service interface
+var _ job.Service = &RegistrySynchronizer{}
+
+type RegistrySynchronizer struct {
+	contract *keeper_registry_wrapper.KeeperRegistry
+	doneWG   sync.WaitGroup
+	interval time.Duration
+	job      job.Job
+	orm      ORM
+
+	chDone chan struct{}
+
+	utils.StartStopOnce
+}
+
+func (rs *RegistrySynchronizer) Start() error {
+	if !rs.OkayToStart() {
```

Comment:
I added a safer way of doing this where the init/shutdown logic is wrapped in a closure, e.g.:
```go
func (b *broadcaster) Start() error {
	return b.StartOnce("Log broadcaster", func() error {
		go b.awaitInitialSubscribers()
		return nil
	})
}
```
IIRC, using `OkayToStart`/`OkayToStop` is vulnerable to races.